### PR TITLE
Feature/rtl variable costs

### PR DESCRIPTION
**Pull request by peggles2**
This pull request updates our tests to use React Testing Library for Variable Costs instead of Enzyme.

### CODE REVIEW

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 1]
```diff
@@ -1,182 +1,234 @@
...
+  test('has correct initial fields', () => {
+    const knowVariableCost = screen.getByRole('radio', {
+      name: /yes, i know the total of my variable costs per unit/i
+    })
+    const notKnowVariableCost = screen.getByRole('radio', {
+      name: /no, input values individually/i
+    })
+    
+    expect(knowVariableCost.value).toBeNull
+    expect(notKnowVariableCost.value).toBeNull
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:28]: Check for input field and button presence here

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 2]
```diff
@@ -1,182 +1,234 @@
...
+    userEvent.click(screen.getByLabelText(knowVariableCost))
+    userEvent.type(screen.getByLabelText('total fixed cost'), '1000')
+    fireEvent.submit(screen.getByTestId('fixedCosts-form'))
+    userEvent.type(screen.getByRole('spinbutton', { name: /per unit selling price\*/i }), '25')
+    userEvent.click(screen.getByRole('button', { name: /continue/i }))
+    userEvent.type(screen.getByRole('spinbutton', {  name: /number of units to sell\*/i}), '123')
+    userEvent.click(screen.getByRole('button', { name: /continue/i }))
+  })
+
+  test('has correct initial fields', () => {
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:19]: Rename to something more concrete.

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 3]
```diff
@@ -1,182 +1,234 @@
...
+    
+    expect(knowVariableCost.value).toBeNull
+    expect(notKnowVariableCost.value).toBeNull
   })
 
-  it('renders without crashing', () => {
-    const wrapper = shallow(<VariableCosts {...baseProps}/>);
-    expect(wrapper).toHaveLength(1);
-  })
+  test('sets VariableCosts value to true on radio button selection', () => {
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:31]: we may not need this test, it seems in conflict with RTL principles. It's not necessarily checking something a user can see. Up to you though

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 4]
```diff
@@ -1,182 +1,234 @@
...
   })
 
-  it('shows variable cost suggestion box on yes selection', () => {
-    const wrapper = shallow(<VariableCosts {...baseProps}/>);
-    expect(wrapper.find('.variableCost-suggestion')).toHaveLength(0)
-    wrapper.setState({knowVariableCosts: 'yes'})
-    expect(wrapper.find('.variableCost-suggestion')).toHaveLength(1)
-  })
+  test('shows the continue button when a selection has been made', () => {
+    const knowVariableCost = screen.getByRole('radio', {
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:50]: this variable is used in a bunch of tests, maybe pull it out to the top of describe block? Or put it in beforeEach?

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 5]
```diff
@@ -1,182 +1,234 @@
...
-    wrapper.setState({knowVariableCosts: 'no'})
-    expect(wrapper.find('NumbersInputForm')).toHaveLength(1)
+    expect(screen.getByRole('button', { name: /continue/i })).toBeInTheDocument()
+    
+    userEvent.click(notKnowVariableCost)
+  
+    expect(screen.getByRole('button', { name: /continue/i })).toBeInTheDocument()
+  })
+
+  test('shows variable cost suggestion box on yes selection', () => {
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:68]: what's a suggestion box?

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 6]
```diff
@@ -1,182 +1,234 @@
...
+    userEvent.click(screen.getByRole('button', { name: /add variable costs individually/i }))
+  
+    expect(screen.queryByRole('spinbutton', { name: /total monthly variable costs\*/i })).not.toBeInTheDocument()
+    expect(screen.queryByRole('spinbutton', { name: /direct materials/i })).toBeInTheDocument()
+    expect(screen.queryByRole('spinbutton', { name: /piece rate labor/i })).toBeInTheDocument()
+    expect(screen.queryByRole('spinbutton', { name: /production supplies/i })).toBeInTheDocument()
+    expect(screen.queryByRole('spinbutton', { name: /commissions/i })).toBeInTheDocument()
+    expect(screen.queryByRole('spinbutton', { name: /freight out/i })).toBeInTheDocument()
+    expect(screen.queryByRole('spinbutton', { name: /other variable costs/i })).toBeInTheDocument()
+  })
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:99]: 👍

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 7]
```diff
@@ -1,182 +1,234 @@
...
+
+  test('calculates the variable cost value and when you go to the next page it displays it', () => {
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:116]: 👍

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 8]
```diff
@@ -1,182 +1,234 @@
...
+      name: /no, input values individually/i
+    })
+    userEvent.click(notKnowVariableCost)
+    userEvent.type(screen.getByRole('spinbutton', { name: /commissions/i }), '12')
+    userEvent.click(screen.getByRole('button', { name: /continue/i }))
+
+    screen.getByText(/\$12/i)
+  })
+
+  test('saves the variable cost value and when you go to the next page it displays it', () => {
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:127]: Is this test doubled? It's named the same as a test above it

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 9]
```diff
@@ -1,182 +1,234 @@
...
+      name: /yes, i know the total of my variable costs per unit/i
+    })
+    userEvent.click(knowVariableCost)
+    userEvent.type(screen.getByRole('spinbutton', { name: /total monthly variable costs\*/i }), '12')
+    userEvent.click(screen.getByRole('button', { name: /continue/i }))
+
+    expect(screen.getByText(/\$12/i)).toBeInTheDocument()
+  })
+  
+  it('Does not go to the next step when variable cost field is empty', () => {
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:138]: 👏

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 10]
```diff
@@ -1,182 +1,234 @@
...
+
+    expect(screen.getByRole('heading', { name: /do you know your variable cost per unit\?\*/i })).toBeInTheDocument()
+    expect(screen.queryByRole('heading', { name: /break-even point results/i })).not.toBeInTheDocument()
   })
 
-  it('has the correct number of fields in the form', () => {
-    const wrapper = shallow(<VariableCosts {...baseProps}/>);
-    wrapper.setState({knowVariableCosts: 'no'})
-    expect(wrapper.find('NumbersInputForm').dive()).toHaveLength(6)
+  it('resets fields on radio button switch', () => {
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:151]: Very good

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 11]
```diff
@@ -1,182 +1,234 @@
...
+    userEvent.click(screen.getByRole('button', { name: /continue/i }))
+    expect(screen.getByText(/Enter a valid variable cost per unit to continue/i)).toBeInTheDocument()
   })
 
-  it('calls setVariableCost and goToStep functions on submit', () => {
-    const setVariableCostMock = jest.fn()
-    const wrapper = shallow(
-      <VariableCosts {...baseProps} setVariableCost={setVariableCostMock} goToStep={goToStepMock} />
-    );
+  test('displays warning message when appropriate', () => {
```
Comment (kscoulter) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:211]: rename displays warning message when variable cost is higher than price per unit

**Review by kscoulter — CHANGES_REQUESTED** [review]

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js (hunk 1 above)
Comment (peggles2) [reply solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:28]: This is just checking the initial state before any radio buttons are clicked. So input fields are not displaying. That is checked in the other tests.

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js (hunk 3 above)
Comment (peggles2) [reply solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:31]: it testes that when you click on the radio button the button is selected. Good to test in case both buttons get selected or none get selected etc.

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js (hunk 5 above)
Comment (peggles2) [reply solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:68]: Screen Shot 2021-01-29 at 3 19 34 PM

Comment (peggles2) [reply solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:68]: Screen Shot 2021-01-29 at 3 23 05 PM

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js (hunk 4 above)
Comment (peggles2) [reply solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:50]: 👍  good idea

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js (hunk 8 above)
Comment (peggles2) [reply solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:127]: One is for individual vc the other is when you know your total. I'll update the description

File: solution/ui/src/components/organisms/fixedCosts/fixedCosts.test.js [hunk 12]
```diff
@@ -15,21 +15,25 @@ describe('FixedCosts', () => {
...
+    knowFixedCost = screen.getByRole('radio', { 
+      name: /yes, i know the total of my monthly fixed costs/i 
+    })
+    notKnowFixedCost = screen.getByRole('radio', { 
+      name: /no, input values individually/i 
+    })
   });
 
   test('Has a heading called "Calculate your total fixed costs"', () => {
     screen.getByRole('heading', { name: /calculate your total fixed costs/i })
```
Comment (mikewolfd) [inline solution/ui/src/components/organisms/fixedCosts/fixedCosts.test.js:27]: are you missing an expect statement here?

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 13]
```diff
@@ -1,182 +1,186 @@
...
-      <VariableCosts {...baseProps} setVariableCost={setVariableCostMock} goToStep={goToStepMock} />
-    );
-
-    wrapper.setState({knowVariableCosts: 'yes'})
-    wrapper.find('Form').simulate('submit')
+    expect(screen.getByRole('button', { name: /continue/i })).toBeInTheDocument()
+    
+    userEvent.click(notKnowVariableCost)
+  
+    expect(screen.getByRole('button', { name: /continue/i })).toBeInTheDocument()
```
Comment (mikewolfd) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:54]: this button should still be here after the previous click? or are you missing .not here?

File: solution/ui/src/components/organisms/fixedCosts/fixedCosts.test.js (hunk 12 above)
Comment (mikewolfd) [inline solution/ui/src/components/organisms/fixedCosts/fixedCosts.test.js:27]: are you missing an expect check here, otherwise this test is empty?

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 14]
```diff
@@ -1,182 +1,186 @@
...
+  })
+
+  test('calculates the variable cost values individually and displays total on next page', () => {
+    userEvent.click(notKnowVariableCost)
+    userEvent.type(screen.getByRole('spinbutton', { name: /commissions/i }), '12')
+    userEvent.type(screen.getByRole('spinbutton', { name: /direct materials/i }), '5')
+
+    userEvent.click(screen.getByRole('button', { name: /continue/i }))
+
+    screen.getByText(/\$17/i)
```
Comment (mikewolfd) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:102]: expect statement?

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js [hunk 15]
```diff
@@ -1,182 +1,186 @@
...
-    const setVariableCostMock = jest.fn()
-    const wrapper = shallow(
-      <VariableCosts {...baseProps} setVariableCost={setVariableCostMock} goToStep={goToStepMock} />
-    );
-    wrapper.setState({knowVariableCosts: 'no'})
-    wrapper.find('Form').simulate('submit')
-
-    expect(goToStepMock).toHaveBeenCalledTimes(0)
-    expect(wrapper.state().formError).toEqual(true)
+    expect(screen.getByText(warningMsg)).toBeInTheDocument
```
Comment (mikewolfd) [inline solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:175]: toBeInTheDocument()

**Review by mikewolfd — CHANGES_REQUESTED** [review]
There seems to be some remaining issues, I left the appropriate comments

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js (hunk 13 above)
Comment (peggles2) [reply solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:54]: either radio button selected the button should be there. It's only not there when nothing is selected.

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js (hunk 15 above)
Comment (peggles2) [reply solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:175]: good catch!

File: solution/ui/src/components/organisms/variableCosts/variableCosts.test.js (hunk 14 above)
Comment (peggles2) [reply solution/ui/src/components/organisms/variableCosts/variableCosts.test.js:102]: the test works anyway but I agree expect statements make it more readable so I'll update the code thanks!

**Review by mikewolfd — APPROVED** [review]
Looks good!
